sys/psa_crypto/psa_crypto.c

```diff
@@ -0,0 +1,1747 @@
+                                    size_t *iv_length)
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    if (!lib_initialized) {
+        return PSA_ERROR_BAD_STATE;
+    }
+
+    *iv_length = 0;
+
+    if (!operation->iv_required || operation->iv_set) {
+        return PSA_ERROR_BAD_STATE;
+    }
+
+    if (iv_size < operation->default_iv_length) {
+        psa_cipher_abort(operation);
+        return PSA_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    status = psa_generate_random(iv, iv_size);
+    if (status != PSA_SUCCESS) {
+        return status;
+    }
+
+    operation->iv_set = 1;
+    *iv_length = operation->default_iv_length;
+
+    return status;
+}
+
+psa_status_t psa_cipher_set_iv(psa_cipher_operation_t *operation,
+                               const uint8_t *iv,
+                               size_t iv_length)
+{
+    (void)operation;
+    (void)iv;
+    (void)iv_length;
+    return PSA_ERROR_NOT_SUPPORTED;
+}
+
+psa_status_t psa_cipher_update(psa_cipher_operation_t *operation,
+                               const uint8_t *input,
+                               size_t input_length,
+                               uint8_t *output,
+                               size_t output_size,
+                               size_t *output_length)
+{
+    (void)operation;
+    (void)input;
+    (void)input_length;
+    (void)output;
+    (void)output_size;
+    (void)output_length;
+    return PSA_ERROR_NOT_SUPPORTED;
+}
+
+
+
+psa_status_t psa_hash_setup(psa_hash_operation_t *operation,
+                            psa_algorithm_t alg)
+{
+    if (!lib_initialized) {
+        return PSA_ERROR_BAD_STATE;
+    }
+
+    if (operation->alg != 0) {
+        return PSA_ERROR_BAD_STATE;
+    }
+
+    if (!PSA_ALG_IS_HASH(alg)) {
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+    psa_status_t status = psa_algorithm_dispatch_hash_setup(operation, alg);
+    if (status == PSA_SUCCESS) {
+        operation->alg = alg;
+    }
+
+    return status;
+}
+
+psa_status_t psa_hash_update(psa_hash_operation_t *operation,
+                             const uint8_t *input,
+                             size_t input_length)
+{
+    if (!lib_initialized) {
+        return PSA_ERROR_BAD_STATE;
+    }
+
+    if (operation->alg == 0) {
+        return PSA_ERROR_BAD_STATE;
+    }
+
+    psa_status_t status = psa_algorithm_dispatch_hash_update(operation, input, input_length);
+
+    if (status != PSA_SUCCESS) {
+        psa_hash_abort(operation);
+    }
+    return status;
+}
+
+psa_status_t psa_hash_finish(psa_hash_operation_t *operation,
+                             uint8_t *hash,
+                             size_t hash_size,
+                             size_t *hash_length)
+{
+    if (!lib_initialized) {
+        return PSA_ERROR_BAD_STATE;
+    }
+
+    if (operation->alg == 0) {
+        return PSA_ERROR_BAD_STATE;
+    }
+
+    uint8_t actual_hash_length = PSA_HASH_LENGTH(operation->alg);
+
+    if (hash_size < actual_hash_length) {
```

Comment:
We actually need the check in `psa_hash_finish`, in case it is called outside of `psa_hash_compute`.
I do agree that it would be good to abort before computing anything, but checking in both places would result in redundant code. I'm not sure what's preferable here.